HPCC-15375 Add default feature level authorization

rpastrana opened the pull request:
- Adds ESPService attribute "requiredaccess"
  If requiredaccess is not declared, build process will fail
  with message to developer
  All methods will require authorization of the feature:level specified
- Adds ESPmethod attribute "requiredaccess" which can overwrite service
  level requiredaccess
- User can specify requiredaccss levels: (NONE|INMETHOD|ACCESS|READ|WRITE|FULL)
  Default is READ

Signed-off-by: rpastrana rodrigo.pastrana@lexisnexis.com

rpastrana committed:
HPCC-15375 Add default feature level authorization

- Adds ESPService attribute "requiredaccess"
   If requiredaccess is not declared, build process will fail
   with message to developer
   All methods will require authorization of the feature:level specified

- Adds ESPmethod attribute "requiredaccess" which can overwrite service
    level requiredaccess
- User can specify requiredaccss levels: (NONE|INMETHOD|ACCESS|READ|WRITE|FULL)
   Default is READ

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@afishbeck please take a look.  The new ESDL flag is currently set to "requiredaccess" but if you had something better in mind, let me know. Ignore the ECM files for all services execp wsesdlconfig and wsworkunits.

HPCCSmoketest commented:
Automated Smoketest 
Sha: ea4117e354904b9be15f3cba8609e39d84dec1af
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on esp/platform/sechandler.cpp:
```diff
@@ -163,6 +163,48 @@ bool SecHandler::authorizeSecFeatures(StringArray & features, IEspStringIntMap &
+
+        const char * key = (const char *)cur.getKey();
+        SecAccessFlags val = *accessmap.getValue(key);
+        features.append((const char *)cur.getKey());
```
Just use 'key' variable from above?

afishbeck commented on esp/platform/sechandler.cpp:
```diff
@@ -163,6 +163,48 @@ bool SecHandler::authorizeSecFeatures(StringArray & features, IEspStringIntMap &
+        const char * key = (const char *)cur.getKey();
+        SecAccessFlags val = *accessmap.getValue(key);
+        features.append((const char *)cur.getKey());
+        reqarray[index++] = *accessmap.getValue(key);
```
just use 'val' from above?

afishbeck commented on esp/platform/sechandler.cpp:
```diff
@@ -163,6 +163,48 @@ bool SecHandler::authorizeSecFeatures(StringArray & features, IEspStringIntMap &
+        reqarray[index++] = *accessmap.getValue(key);
+
+        DBGLOG("ITERATING: key: %s val: %d", key, val);
+        DBGLOG("reqarray[%d]: val: %d", index-1, reqarray[index-1]);
```
Logging seems pretty verbose for building an array.

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5373,57 +5374,166 @@ void EspServInfo::write_factory_impl()
+        return "SecAccess_Access";
+
+    //we might need to throw here...
+    outs(2, "\n//FEATURE LEVEL VALUE INVALID: DEFAULTING REQUIRED LEVEL to 'READ'!\n//Valid values are NONE INMETHOD, ACCESS, READ, WRITE, FULL\n");
```
Better to be strict and avoid user error?

rpastrana replied:
Agreed, I had FULL as the default, but noticed that somewhere in the secmanager "READ" is the default....

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5373,57 +5374,166 @@ void EspServInfo::write_factory_impl()
+        return "SecAccess_None";
+    }
+
+    if (strieq(level, "INMETHOD"))
```
Not sure I like the term "INMETHOD".  Could be convinced, but what do you think of "DELAYED" or "INLINE"?

rpastrana replied:
I'm not crazy about any of them actually... my intent was to allow developers to declare that the feature level authorization logic is to be implemented by the developer (therefore automatic authorization is suppressed)....

Maybe "DEFERRED"  makes sense if the attribute name changes to automaticFeatureAuth or something of the like. Thoughts? @afishbeck

afishbeck commented on esp/scm/WsDeploy.ecm:
```diff
@@ -568,7 +568,7 @@ ESPresponse [exceptions_inline, encode(0)] GetSummaryResponse
 };
 
-ESPservice [exceptions_inline("xslt/exceptions.xslt")] WsDeploy
+ESPservice [requiredaccess("NONE"), exceptions_inline("xslt/exceptions.xslt")] WsDeploy
```
Saying NONE at this level is kind of confusing, because in fact it still does service level authentication, right?  That is it still prompts me for a valid user id and password?  Somehow seems like we need to make clear the difference between accessing the process, the service, and the features.  It may be as simple as renaming this attribute.

It currently sounds like "service required access" but you are really talking about "default feature level access".

Up for discussion, but maybe something like everywhere that you currently have "requiredaccess", change to "featureaccess" to be less confusing.

rpastrana replied:
The "NONE" keyword makes more sense if the attribute becomes more explicit (and accurate)... above I threw out a very verbose option of "autoFeatureAuth" which I like because it is self documenting, I'll likely implement this change unless you have a strong opinion against it... @afishbeck

afishbeck commented:
@rpastrana made some general comments for discussion.

rpastrana commented:
@afishbeck please take a look.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 8340beb796e565a6e382bbdf35453baa6b20a96f
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@rpastrana  Those (minor seeming) changes do make it less confusing.  "autofeatureauth" seems a bit long compared to just "featureauth" but it's fine if you think it makes some aspect more clear.

Besides a slightly long tag name looks good.

rpastrana committed:
HPCC-15375 Code review changes

- Change INMETHOD keyword to DEFERRED
- Change requiredaccess attribute to autofeatureauth
- Change default access level to FULL

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest 
Sha: c86746baaa530e6eb1a0e52735d548df73f118d6
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@afishbeck @wangkx @RussWhitehead this change adds automatic Feature Level Auth to all ESP Methods declared in ESDL (ECM) files.
I will be asking ESP service "owners" to verify the new behavior and to advise on new approach if a given method is found to lack feature level auth.

rpastrana commented on esp/scm/WsDeploy.ecm:
```diff
@@ -568,7 +568,7 @@ ESPresponse [exceptions_inline, encode(0)] GetSummaryResponse
 };
 
-ESPservice [requiredaccess("NONE"), exceptions_inline("xslt/exceptions.xslt")] WsDeploy
+ESPservice [autofeatureauth("NONE"), exceptions_inline("xslt/exceptions.xslt")] WsDeploy
```
@wangkx none of the methods in this service seem to require any feature level auth, is this correct?

wangkx replied:
@garonsky is WsDeploy used by ConfigMgr?

rpastrana replied:
@garonsky I think we've talked about this before, but please confirm if it makes sense for wsdeploy not to perform feature level auth... thanks.

rpastrana commented on esp/scm/esploggingservice.ecm:
```diff
@@ -115,7 +115,7 @@ TransactionSeedResponse
 };
 
-ESPService [requiredaccess("NONE"),noforms] WsLogService
+ESPService [autofeatureauth("NONE"),noforms] WsLogService
```
@wangkx does it make sense to perform feature level auth on any/all of these methods?

wangkx replied:
For now, I think that feature level auth should be on all of methods.

rpastrana replied:
@wangkx what feature name, and level?

wangkx replied:
New feature name and WRITE

rpastrana replied:
EspLoggingAccess? I didn't see it in the buildset.xml, I'll open a new subtask.

rpastrana replied:
https://track.hpccsystems.com/browse/HPCC-16312

rpastrana commented on esp/scm/ws_access.ecm:
```diff
@@ -787,8 +787,8 @@ ESPresponse [nil_remove] UserAccountExportResponse
-
-ESPservice [requiredaccess("NONE"), version("1.09"), exceptions_inline("./smc_xslt/exceptions.xslt")] ws_access
+//Kevin/russ does this method need feature level check?
+ESPservice [autofeatureauth("NONE"), version("1.09"), exceptions_inline("./smc_xslt/exceptions.xslt")] ws_access
```
@wangkx @RussWhitehead are we intentionally not requiring feature level auth on ws_access' methods?

wangkx replied:
I think we should do feature level auth on ws_access' methods.

rpastrana replied:
https://track.hpccsystems.com/browse/HPCC-16313

rpastrana commented on esp/scm/ws_account.ecm:
```diff
@@ -67,7 +67,8 @@ ESPresponse [exceptions_inline] VerifyUserResponse
 
-ESPservice [requiredaccess("NONE"), version("1.01"), default_client_version("1.01"), exceptions_inline("./smc_xslt/exceptions.xslt")] ws_account
+//Kevin/russ does this method need feature level check?
+ESPservice [autofeatureauth("NONE"), version("1.01"), default_client_version("1.01"), exceptions_inline("./smc_xslt/exceptions.xslt")] ws_account
```
@wangkx @RussWhitehead are we intentionally not requiring feature level auth on ws_account's methods?

wangkx replied:
The ws_account is for the user who logged in. So, we do not need feature level auth.

rpastrana replied:
Sounds good, I'll leave as 'NONE'

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -729,15 +729,19 @@ ESPservice [
     ESPmethod [resp_xsl_default("/esp/xslt/dfu_viewdata.xslt")] DFUBrowseData(DFUBrowseDataRequest, DFUBrowseDataResponse);
     ESPmethod [resp_xsl_default("/esp/xslt/dfu_searchdata.xslt")] DFUSearchData(DFUSearchDataRequest, DFUSearchDataResponse);
+    //Kevin I didn't see feature level check for DFUGetFileMetaData
     ESPmethod DFUGetFileMetaData(DFUGetFileMetaDataRequest, DFUGetFileMetaDataResponse);
```
@wangkx I didn't see feature level check for DFUGetFileMetaData

wangkx replied:
Should be added.

rpastrana replied:
@wangkx  what feature and level?

wangkx replied:
DfuAccess READ

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -729,15 +729,19 @@ ESPservice [
     ESPmethod DFUDefFile(DFUDefFileRequest, DFUDefFileResponse);
 
     ESPmethod [resp_xsl_default("/esp/xslt/addto_superfile.xslt")] AddtoSuperfile(AddtoSuperfileRequest, AddtoSuperfileResponse);
+    //Kevin I didn't see feature level check for SuperfileList
```
@wangkx can we confirm we don't need feature level auth for the methods I commented on below?

wangkx replied:
Should be added.

rpastrana replied:
@wangkx  feature and level?

wangkx replied:
DfuAccess READ

rpastrana commented on esp/scm/ws_dfuXref.ecm:
```diff
@@ -232,10 +232,12 @@ ESPservice [requiredaccess("NONE"), exceptions_inline("./smc_xslt/exceptions.xsl
     ESPmethod [resp_xsl_default("/esp/xslt/xref_directories.xslt")] DFUXRefDirectories(DFUXRefDirectoriesQueryRequest, DFUXRefDirectoriesQueryResponse);
-    ESPmethod [resp_xsl_default("/esp/xslt/action.xslt")] DFUXRefArrayAction(DFUXRefArrayActionRequest, DFUXRefArrayActionResponse);
+    //Tony/kevin, this method is not protected, should it be?
+    ESPmethod [autofeatureauth("DfuXrefAccess:access"), resp_xsl_default("/esp/xslt/action.xslt")] DFUXRefArrayAction(DFUXRefArrayActionRequest, DFUXRefArrayActionResponse);
```
@wangkx @afishbeck this method is not protected, should it be?

wangkx replied:
Should be added.

rpastrana commented on esp/scm/ws_dfuXref.ecm:
```diff
@@ -232,10 +232,12 @@ ESPservice [requiredaccess("NONE"), exceptions_inline("./smc_xslt/exceptions.xsl
     ESPmethod DFUXRefCleanDirectories(DFUXRefCleanDirectoriesRequest, DFUXRefCleanDirectoriesResponse);
-    ESPmethod DFUXRefUnusedFiles(DFUXRefUnusedFilesRequest, DFUXRefUnusedFilesResponse);
+    //Tony/kevin, this method is not protected, should it be?
+    ESPmethod [autofeatureauth("DfuXrefAccess:access")] DFUXRefUnusedFiles(DFUXRefUnusedFilesRequest, DFUXRefUnusedFilesResponse);
```
@wangkx @afishbeck this method is not protected, should it be?

wangkx replied:
Should be added.

rpastrana commented on esp/scm/ws_espcontrol.ecm:
```diff
@@ -30,7 +30,8 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] SetLoggingResponse
 };
 
-ESPservice [ requiredaccess("NONE"), version("1.00"), default_client_version("1.00"), exceptions_inline("./smc_xslt/exceptions.xslt")] WSESPControl
+//no authfeature is performed, however in cpp, there's a check to see if user is superuser
```
@RussWhitehead is the superuser check enough?

rpastrana replied:
@RussWhitehead please confirm this is the correct behavior

RussWhitehead replied:
This is correct

rpastrana commented on esp/scm/ws_packageprocess.ecm:
```diff
@@ -288,7 +288,7 @@ ESPresponse [exceptions_inline] GetPartFromPackageMapResponse
 
-ESPservice [requiredaccess("NONE"), version("1.02"), exceptions_inline("./smc_xslt/exceptions.xslt")] WsPackageProcess
+ESPservice [autofeatureauth("NONE"), version("1.02"), exceptions_inline("./smc_xslt/exceptions.xslt")] WsPackageProcess
 {
```
@wangkx @afishbeck it looks like this entire service lacks feature level auth, can you guys confirm this was done by design?

wangkx replied:
Should be added.

afishbeck replied:
@rpastrana If you add one, will it default to allow, or will operations will have to manually update ldap for any updated environment?  If the latter are you going to coordinate with operations to enable it before they upgrade?

rpastrana replied:
@afishbeck adding new features will be handled separately as sub-tasks. It should be handled by the service "owner", and you're right the new behavior needs to be coordinated with ops and perhaps other users.

rpastrana replied:
https://track.hpccsystems.com/browse/HPCC-16311

rpastrana commented on esp/scm/ws_smc.ecm:
```diff
@@ -397,37 +397,37 @@ ESPresponse [exceptions_inline] LockQueryResponse
 //this one doesn't seem to do any feature level check, Kevin/Tony?
-    ESPmethod [requiredaccess("NONE")] SetJobPriority(SMCPriorityRequest, SMCPriorityResponse);
-    ESPmethod [requiredaccess("INMETHOD")] GetThorQueueAvailability(GetThorQueueAvailabilityRequest, GetThorQueueAvailabilityResponse);
+    ESPmethod [autofeatureauth("NONE")] SetJobPriority(SMCPriorityRequest, SMCPriorityResponse);
```
@wangkx several of these methods don't perform feature level auth, can you confirm? if you're not directly involved w those methods, please let me know.

wangkx replied:
Should be added.

rpastrana replied:
@wangkx smcaccess:WRITE?

wangkx replied:
ok

rpastrana commented on esp/scm/ws_workunits.ecm:
```diff
@@ -1793,9 +1793,8 @@ ESPservice [
-
-    ESPmethod [requiredaccess("OwnWorkunitsAccess:WRITE")] WUCreate(WUCreateRequest, WUCreateResponse);
-    ESPmethod [requiredaccess("OwnWorkunitsAccess:WRITE")] WUCreateAndUpdate(WUUpdateRequest, WUUpdateResponse);
+    ESPmethod [autofeatureauth("OwnWorkunitsAccess:WRITE")] WUCreate(WUCreateRequest, WUCreateResponse);
```
@wangkx I moved the feature level check here from the method implementation, we can revert if you're not comfortable with this change.

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -729,15 +729,19 @@ ESPservice [
     ESPmethod [resp_xsl_default("/esp/xslt/dfu_superedit.xslt")] SuperfileList(SuperfileListRequest, SuperfileListResponse);
     ESPmethod [resp_xsl_default("/esp/xslt/dfu_superresult.xslt")] SuperfileAction(SuperfileActionRequest, SuperfileActionResponse);
-    
+    //Kevin I didn't see feature level check for savexml
```
@wangkx what about Savexml, Add, and AddRemote?

wangkx replied:
Savexml: The name is misleading. The method is used to retrieve file xml (download/'save' to user's local PC). So, DfuAccess: READ.
Add: DfuAccess: WRITE.
AddRemote: DfuAccess: WRITE.

rpastrana replied:
ok

rpastrana commented on esp/scm/ws_smc.ecm:
```diff
@@ -397,37 +397,37 @@ ESPresponse [exceptions_inline] LockQueryResponse
+    ESPmethod GetThorQueueAvailability(GetThorQueueAvailabilityRequest, GetThorQueueAvailabilityResponse);
 //this one ensures superuser, but no feature check is done, Kevin/Tony?
-    ESPmethod [requiredaccess("INMETHOD")] SetBanner(SetBannerRequest, SetBannerResponse);
+    ESPmethod [autofeatureauth("NONE")] SetBanner(SetBannerRequest, SetBannerResponse);
```
@wangkx smcaccess:WRITE?

wangkx replied:
ok

rpastrana commented on esp/scm/ws_smc.ecm:
```diff
@@ -397,37 +397,37 @@ ESPresponse [exceptions_inline] LockQueryResponse
+    ESPmethod [autofeatureauth("NONE")] SetBanner(SetBannerRequest, SetBannerResponse);
 
-    ESPmethod [requiredaccess("NONE")] NotInCommunityEdition(NotInCommunityEditionRequest, NotInCommunityEditionResponse);
+    ESPmethod [autofeatureauth("NONE")] NotInCommunityEdition(NotInCommunityEditionRequest, NotInCommunityEditionResponse);
```
@wangkx smcaccess:READ?

wangkx replied:
ok

rpastrana commented on esp/scm/ws_smc.ecm:
```diff
@@ -397,37 +397,37 @@ ESPresponse [exceptions_inline] LockQueryResponse
+    ESPmethod RoxieControlCmd(RoxieControlCmdRequest, RoxieControlCmdResponse);
 //this one doesn't seem to do any feature level check, Kevin/Tony?
-    ESPmethod [requiredaccess("NONE")] GetStatusServerInfo(GetStatusServerInfoRequest, GetStatusServerInfoResponse);
+    ESPmethod [autofeatureauth("NONE")] GetStatusServerInfo(GetStatusServerInfoRequest, GetStatusServerInfoResponse);
```
@wangkx smcaccess:READ?

wangkx replied:
ok

rpastrana commented on esp/scm/ws_smc.ecm:
```diff
@@ -397,37 +397,37 @@ ESPresponse [exceptions_inline] LockQueryResponse
+    ESPmethod [autofeatureauth("NONE")] GetStatusServerInfo(GetStatusServerInfoRequest, GetStatusServerInfoResponse);
 //this one doesn't seem to do any feature level check, Kevin/Tony?
-    ESPmethod LockQuery(LockQueryRequest, LockQueryResponse);
+    ESPmethod [autofeatureauth("NONE")] LockQuery(LockQueryRequest, LockQueryResponse);
```
@wangkx smcaccess:WRITE?

wangkx replied:
ok

rpastrana commented on esp/scm/ws_espcontrol.ecm:
```diff
@@ -30,7 +30,8 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] SetLoggingResponse
 };
 
-ESPservice [ version("1.00"), default_client_version("1.00"), exceptions_inline("./smc_xslt/exceptions.xslt")] WSESPControl
+//no authfeature is performed, however in cpp, there's a check to see if user is superuser
```
@wangkx @afishbeck Ws_espcontrol::Setlogging prob doesn't need feature level check, but would like confirmation from somebody with intimate knowledge.

wangkx replied:
The method does check whether a user is super user or not.

rpastrana replied:
Yes, that's what I commented, ask for confirmation that the behavior is correct... I'll leave as is.

rpastrana commented:
@wangkx please review ECM files (other logic has been reviewed).

HPCCSmoketest commented:
Automated Smoketest 
Sha: cd61a059c0ad85d086abe6669ee86617476f4e9b
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

rpastrana committed:
HPCC-15375 Changed default auth based on comments from service owners

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest 
Sha: f41ed64593a8ee6028956a6f3520129ab5f4b939
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

afishbeck reviewed (changes requested):
@gaojx please review this pull request.  The functionality here and anything added HPCC-15608 should correspond.  This is for static c++ ESDL services, the latter is for Dynamic ESDL services, but they should have the same syntax and behavior.

rpastrana committed:
HPCC-15375 DESDL Auto ESP feature auth

- change attribute name.

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest 
Sha: 0fdf4910ed3410de92fb3593d4d5baab1c3549a5
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
Needs rebasing for a clean merge

wangkx commented on esp/scm/ws_dfu.ecm:
```diff
@@ -729,20 +729,15 @@ ESPservice [
     ESPmethod [resp_xsl_default("/esp/xslt/dfu_searchdata.xslt")] DFUSearchData(DFUSearchDataRequest, DFUSearchDataResponse);
-    //Kevin I didn't see feature level check for DFUGetFileMetaData
-    ESPmethod DFUGetFileMetaData(DFUGetFileMetaDataRequest, DFUGetFileMetaDataResponse);
+    ESPmethod [autofeatureauth("DfuAccess:READ")]DFUGetFileMetaData(DFUGetFileMetaDataRequest, DFUGetFileMetaDataResponse);
```
@rpastrana do you want to allow an error message being specified when the autofeatureauth fails?

rpastrana replied:
It will be somewhat cumbersome to add syntax for custom error messages, but will considering expanding the current error message to report all the feature level names and their level.

richardkchapman commented:
Needs rebasing for a clean merge

HPCCSmoketest commented:
Automated Smoketest 
Sha: 6d2c0a5f853519ef4edb6a263b8bdfc571321b50
Build: failed
Number of Error(s): 3
 ESDL Error: ws_access service definition must declare default feature access. Example \ESPservice [auth_feature(MyServiceAccess:FULL)]\
/mnt/disk1/home/vamosax/smoketest/PR-9063/build/generated/ws_access.esp:27638:1: error: \xe2\x80\x98ESDL\xe2\x80\x99 does not name a type
 ESDL Error: ws_access service definition must declare default feature access. Example \ESPservice [auth_feature(MyServiceAccess:FULL)]\

rpastrana committed:
HPCC-15375 Report all feature levels need upon failure

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please review latest changes.

wangkx commented:
@rpastrana  the changes look fine.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 16e8874b697c07b900ceeb4022069900644be068
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@richardkchapman this change is ready to be merged.
